File: ios/Sources/Omnesis/Photos/PhotosSettings.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

import Foundation

/// Per-user Photos source preferences. Lives in `UserDefaults.standard`
/// so it survives app relaunches but is wiped on uninstall. Mirrors
/// `HealthSettings`'s shape (same `KeyValueDefaults` seam for testing).
public struct PhotosSettings: Sendable {
    public enum Keys {
        /// User has opted in to the Photos source (post-pairing flow).
        /// Drives whether the gateway registers a `photos:<account>`
        /// source row + whether the collector instantiates `PhotosSource`.
        public static let enabled = "omnesis.photos.enabled"
        /// ISO-8601 timestamp of when the source was last enabled.
        /// Recorded for display/diagnostics; rich-analysis eligibility is
        /// decided entirely by `AnalyzedAssetStore` (has this asset been
        /// richly analyzed yet), not by this watermark.
        public static let enabledAt = "omnesis.photos.enabledAt"
        /// ISO-8601 timestamp of the last whole-library reconcile pass
        /// (the deletion backstop). Throttled to at most once/day — see
        /// `AppStore.runPhotosBackfillTask`, the sole reader/writer of
        /// this key.
        public static let lastReconcileAt = "omnesis.photos.lastReconcileAt"
        public static let accessState = "omnesis.photos.accessState"
        public static let accessEpoch = "omnesis.photos.accessEpoch"
    }

    private let defaults: KeyValueDefaults

    public init(defaults: KeyValueDefaults = UserDefaults.standard) {
        self.defaults = defaults
    }

    public var enabled: Bool {
        get { (defaults.object(forKey: Keys.enabled) as? Bool) ?? false }
        nonmutating set { defaults.set(newValue, forKey: Keys.enabled) }
    }

    public var enabledAt: Date? {
        get {
            (defaults.string(forKey: Keys.enabledAt)).flatMap { ISO8601DateFormatter().date(from: $0) }
        }
        nonmutating set {
            if let newValue {
                defaults.set(ISO8601DateFormatter().string(from: newValue), forKey: Keys.enabledAt)
            } else {
                defaults.removeObject(forKey: Keys.enabledAt)
            }
        }
    }

    public var lastReconcileAt: Date? {
        get {
            (defaults.string(forKey: Keys.lastReconcileAt)).flatMap { ISO8601DateFormatter().date(from: $0) }
        }
        nonmutating set {
            if let newValue {
                defaults.set(ISO8601DateFormatter().string(from: newValue), forKey: Keys.lastReconcileAt)
            } else {
                defaults.removeObject(forKey: Keys.lastReconcileAt)
            }
        }
    }

    public var lastAccessState: PhotosAccessState? {
        get { defaults.string(forKey: Keys.accessState).flatMap(PhotosAccessState.init(rawValue:)) }
        nonmutating set {
            if let newValue {
                defaults.set(newValue.rawValue, forKey: Keys.accessState)
            } else {
                defaults.removeObject(forKey: Keys.accessState)
            }
        }
    }

    public var accessEpoch: Int {
        get { (defaults.object(forKey: Keys.accessEpoch) as? Int) ?? 0 }
        nonmutating set { defaults.set(newValue, forKey: Keys.accessEpoch) }
    }

    /// Persist one observed authorization transition. Returning true means a
    /// previously scoped/inaccessible library became complete and the source
    /// must replay its full baseline under the new epoch.
    @discardableResult
    public func observeAccess(_ current: PhotosAccessState) -> Bool {
        // A missing state is also an upgrade boundary. Existing installs can
        // already have a steady epoch-0 cursor created before access epochs
        // existed, so the first observed Full grant must force one safe replay.
        let restored = current.isComplete && (lastAccessState?.isComplete != true)
        if restored {
            accessEpoch += 1
        }
        lastAccessState = current
        return restored
    }

    public func recordReconcile(_ outcome: PhotosSource.ReconcileOutcome, at date: Date = Date()) {
        guard outcome == .performed else { return }
        lastReconcileAt = date
    }

    /// Enable the source, stamping `enabledAt` only the FIRST time (a
    /// disable/re-enable cycle keeps the original watermark).
    public func enable(now: Date = Date()) {
        // Detaching a partition removes its gateway history, while the local
        // analysis index survives. Replay that index under a fresh generation
        // on re-enable, including when Photos access remains Limited.
        if !enabled {
            accessEpoch += 1
            lastReconcileAt = nil
        }
        enabled = true
        if enabledAt == nil {
            enabledAt = now
        }
    }

    public func disable() {
        enabled = false
    }

    /// Wipe pairing-scoped settings. The access epoch is intentionally
    /// monotonic across unpair/re-pair because the on-device analyzed-asset
    /// index also survives unpair; resetting only one side would make records
    /// from a higher epoch impossible to mark current after re-pairing.
    public func reset() {
        defaults.removeObject(forKey: Keys.enabled)
        defaults.removeObject(forKey: Keys.enabledAt)
        defaults.removeObject(forKey: Keys.lastReconcileAt)
        defaults.removeObject(forKey: Keys.accessState)
    }
}

/// Orders a Photos authorization callback so a restored Full grant advances
/// the cursor epoch and rebuilds the collector before health is published and
/// syncing resumes.
@available(iOS 17.0, *)
@MainActor
enum PhotosPermissionTransitionCoordinator {
    static func authorizationChanged(
        access: PhotosAccessState,
        settings: PhotosSettings,
        rebuild: () async -> Void,
        refresh: () async -> Void,
        sync: () async -> Void
    ) async {
        let previous = settings.lastAccessState
        if settings.observeAccess(access) {
            await rebuild()
        }
        if previous != access {
            await refresh()
        }
        await sync()
    }
}

File: android/feature-photos/src/main/kotlin/dev/omnesis/android/feature/photos/PhotosSettings.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

package dev.omnesis.android.feature.photos

import dev.omnesis.android.transport.PushRefusalLedger
import java.time.Instant

/**
 * Abstract string key/value store backing [PhotosSettings]. SharedPreferences
 * conforms in production (see `di/PhotosModule`); tests inject an
 * [InMemoryKeyValueStore].
 */
interface KeyValueStore {
    fun get(key: String): String?

    /** Stores [value] under [key]; null removes the entry. */
    fun put(key: String, value: String?)
}

/** In-memory backing for tests. */
class InMemoryKeyValueStore(
    initial: Map<String, String> = emptyMap(),
) : KeyValueStore {
    private val values = initial.toMutableMap()

    override fun get(key: String): String? = synchronized(values) { values[key] }

    override fun put(key: String, value: String?) {
        synchronized(values) {
            if (value == null) values.remove(key) else values[key] = value
        }
    }
}

/**
 * Per-user Photos preferences. Survives app relaunches (SharedPreferences in
 * production) but is wiped on uninstall.
 *
 * Two pieces of state:
 *  1. [photosEnabled] — the user has opted in to Photos syncing. Drives
 *     whether the gateway registers a `photos:local` source row and whether
 *     the sync worker + media-observer job run at all.
 *  2. [lastReconcileAt] — when the whole-library reconcile snapshot last
 *     ran. Throttled to at most once/day (see `PhotosSyncCoordinator`) since
 *     a photo library is orders of magnitude larger than, say, a call log —
 *     sending every current asset id on every sync pass would be real cost.
 */
class PhotosSettings(private val store: KeyValueStore) {

    object Keys {
        const val ENABLED = "omnesis.photos.enabled"
        const val PERMISSION_PERMANENTLY_DENIED = "omnesis.photos.permissionPermanentlyDenied"
        const val LAST_RECONCILE_AT = "omnesis.photos.lastReconcileAt"
        const val PUSH_PREFIX = "omnesis.photos.push"
        const val ACCESS = "omnesis.photos.access"
        const val ACCESS_GENERATION = "omnesis.photos.accessGeneration"
    }

    /**
     * Refusals counted against the backfill page the sync is stuck on, and the
     * record of the pages it gave up on. The unit is the cursor position the
     * page started from, which moves only when a page gets through.
     */
    val pushRefusals = PushRefusalLedger(
        sourceId = PhotosSyncCoordinator.SOURCE_ID,
        keyPrefix = Keys.PUSH_PREFIX,
        read = store::get,
        write = store::put,
    )

    var photosEnabled: Boolean
        get() = store.get(Keys.ENABLED) == "true"
        set(value) = store.put(Keys.ENABLED, value.toString())

    var permissionPermanentlyDenied: Boolean
        get() = store.get(Keys.PERMISSION_PERMANENTLY_DENIED) == "true"
        set(value) = store.put(Keys.PERMISSION_PERMANENTLY_DENIED, value.toString())

    fun recordPermissionResult(primaryGranted: Boolean, permanentlyDenied: Boolean) {
        permissionPermanentlyDenied = !primaryGranted && permanentlyDenied
    }

    fun observePermission(primaryGranted: Boolean) {
        if (primaryGranted) permissionPermanentlyDenied = false
    }

    var lastReconcileAt: Instant?
        get() = store.get(Keys.LAST_RECONCILE_AT)?.let { runCatching { Instant.parse(it) }.getOrNull() }
        set(value) = store.put(Keys.LAST_RECONCILE_AT, value?.toString())

    val accessGeneration: Long
        get() = store.get(Keys.ACCESS_GENERATION)?.toLongOrNull() ?: 0L

    /** Monotonic epoch: a LIMITED/DENIED -> FULL expansion requires a complete re-baseline. */
    @Synchronized
    fun observeAccess(access: PhotosAccess): Long {
        val previous = store.get(Keys.ACCESS)?.let { runCatching { PhotosAccess.valueOf(it) }.getOrNull() }
        var generation = accessGeneration
        if (previous != PhotosAccess.FULL && access == PhotosAccess.FULL) {
            generation += 1
            store.put(Keys.ACCESS_GENERATION, generation.toString())
            lastReconcileAt = null
        }
        store.put(Keys.ACCESS, access.name)
        return generation
    }

    /** Wipe every setting — called on unpair so a fresh pairing starts from defaults. */
    fun reset() {
        store.put(Keys.ENABLED, null)
        store.put(Keys.PERMISSION_PERMANENTLY_DENIED, null)
        store.put(Keys.LAST_RECONCILE_AT, null)
        store.put(Keys.ACCESS, null)
        store.put(Keys.ACCESS_GENERATION, null)
        pushRefusals.clear()
    }
}
